engine/include/cubos/engine/cubos.hpp

```diff
@@ -49,44 +50,61 @@ namespace cubos::engine
         /// @tparam F The type of the system function.
         /// @param func The system function.
-        /// @param stage The stage in which the system should be executed.
-        template <typename F> Cubos& addSystem(F func, std::string stage);
+        template <typename F> Cubos& system(F func);
 
         /// Adds a new startup system to the engine.
-        /// Startup systems are executed before the main loop starts.
-        /// If the stage doesn't already exist, it is created in the previously specified default position.
+        /// Startup systems are executed only once, before the main loop starts.
         /// @tparam F The type of the system function.
         /// @param func The system function.
-        /// @param stage The stage in which the system should be executed.
-        template <typename F> Cubos& addStartupSystem(F func, std::string stage);
-
-        /// Sets a given stage to happen after another stage.
-        /// The current stage must exist, but the reference stage may not, in which case it
-        /// will be created.
-        /// @param current The current stage, which will be placed right after the reference stage.
-        /// @param reference The reference stage.
-        Cubos& putStageAfter(const std::string& stage, const std::string& referenceStage);
-
-        /// Sets a given stage to happen before another stage.
-        /// The current stage must exist, but the reference stage may not, in which case it
-        /// will be created.
-        /// @param current The current stage, which will be placed right before the reference stage.
-        /// @param reference The reference stage.
-        Cubos& putStageBefore(const std::string& stage, const std::string& referenceStage);
+        template <typename F> Cubos& startupSystem(F func);
+
+        /// Defines the tag for the current `system`/`startupSystem` set.
+        /// @param tag The tag to run under.
+        Cubos& tagged(const std::string& tag);
+
+        /// Sets the current system to run after the tag.
+        /// If the specified tag doesn't exist, it is internally created.
+        /// @param tag The tag to run after.
+        Cubos& afterTag(const std::string& tag);
+
+        /// Sets the current system to run after a given system.
+        /// The specified system must exist.
+        /// @param system The system to run after.
+        template <typename F> Cubos& after(F system);
+
+        /// Sets the current system to run before the tag.
+        /// If the specified tag doesn't exist, it is internally created.
+        /// @param tag The tag to run before.
+        Cubos& beforeTag(const std::string& tag);
+
+        /// Sets the current system to run before a given system.
+        /// The specified system must exist.
+        /// @param system The system to run before.
+        template <typename F> Cubos& before(F system);
```

Comment:
Prolly a good idea to change it to `beforeSystem` (the same goes for other similar methods).

Reply:
Does this still apply with the builder approach?

Reply:
Yeah, what I mean was that it may the names `afterTag` vs `after` may get confusing since `after` takes a templated argument. So, I think it would be more explicit to rename `after` to `afterSystem`.

Reply:
Ah, yes I agree

Reply:
Implemented